ros/angel_system_nodes/angel_system_nodes/hmm.py

```diff
@@ -0,0 +1,254 @@
+            "query_task_graph",
+            self.query_task_graph_callback
+        )
+
+        # Start the HMM state threads
+        log.info(f"Starting HMM threads")
+        self._hmm_active = Event()
+        self._hmm_active.set()
+        self._hmm_active_heartbeat = 0.1
+
+        self._hmm_state_awake_evt = Event()
+        self._hmm_skip_awake_evt = Event()
+
+        self._hmm_state_thread = Thread(
+            target=self.thread_run_hmm_state,
+            name="hmm_state_runtime"
+        )
+        self._hmm_state_thread.daemon = True
+        self._hmm_state_thread.start()
+
+        self._hmm_skip_thread = Thread(
+            target=self.thread_run_hmm_skip,
+            name="hmm_skip_runtime"
+        )
+        self._hmm_skip_thread.daemon = True
+        self._hmm_skip_thread.start()
+        log.info(f"Starting HMM threads... done")
+
+    def det_callback(self, activity_msg: ActivityDetection):
+        """
+        Callback function for the activity detection subscriber topic.
+        Adds the activity detection msg to the HMM and then publishes a new
+        TaskUpdate message.
+        """
+        if self.hmm_alive():
+            source_stamp_start_frame_sec = time_to_int(
+                activity_msg.source_stamp_start_frame
+            ) * 1e-9 # time_to_int returns ns
+            source_stamp_end_frame_sec = time_to_int(
+                activity_msg.source_stamp_end_frame
+            ) * 1e-9 # time_to_int returns ns
+
+            # Add activity classification to the HMM
+            self._hmm.add_activity_classification(
+                activity_msg.label_vec,
+                activity_msg.conf_vec,
+                source_stamp_start_frame_sec,
+                source_stamp_end_frame_sec,
+            )
+
+            # Tell the HMM threads to wake up
+            self._hmm_state_awake_evt.set()
+            self._hmm_skip_awake_evt.set()
+
+    def publish_task_state_message(self, activity=None):
+        """
+        Forms and sends a `angel_msgs/TaskUpdate` message to the
+        TaskUpdates topic.
+        """
+        log = self.get_logger()
+
+        message = TaskUpdate()
+
+        # Populate message header
+        message.header.stamp = self.get_clock().now().to_msg()
+        message.header.frame_id = "Task message"
+
+        # Populate steps and current step
+        message.steps = self._hmm.model.class_str
+        message.current_step = self._current_step
+        message.current_step_id = self._hmm.model.class_str.index(self._current_step)
+
+        if self._previous_step is None:
+            message.previous_step = "N/A"
+        else:
+            message.previous_step = self._previous_step
+
+        # TODO: Populate task name and description
+        # TODO: Do we need to fill in current/next activity?
+        # TODO: Fill in time remaining?
+
+        self._publisher.publish(message)
+
+        # Update state tracking vars
+        self._previous_step = self._current_step
```

Comment:
The previous and current step may be the same a lot of the time given the current stream prediction context that I'm seeing here. Is that OK? IIRC this was supposed to be the "true" previous step in the task regardless of how long we have been on the current step?